File: 백준/Gold/1753. 최단경로/최단경로.py
import sys
import heapq

readline = sys.stdin.readline
V, E = map(int, readline().split()) # 정점, 간선 개수
K = int(readline().rstrip()) # 시작 정점
edges = [[] for _ in range(V+1)]
dist = [float('inf')] * (V+1)

def dijkstra():
    dist[K] = 0
    hq = []
    heapq.heappush(hq, (0, K))

    while hq:
        now_cost, now_node = heapq.heappop(hq)

        if dist[now_node] < now_cost:
            continue

        for edge in edges[now_node]:
            end_node, weight = edge
            new_cost = now_cost + weight
            if dist[end_node] > new_cost:
                heapq.heappush(hq, (new_cost, end_node))
                dist[end_node] = new_cost

for _ in range(E):
    s, e, w = map(int, readline().split())
    edges[s].append((e, w))


dijkstra()

for v in range(1, V+1):
    if dist[v] == float('inf'):
        print("INF")
    else:
        print(dist[v])

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.*;
import java.util.*;

public class Main {
    static int V, E, K;
    static ArrayList<ArrayList<int[]>> edges;
    static int[] dist;
    static PriorityQueue<int[]> pq;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());

        dist = new int[V+1];
        Arrays.fill(dist, Integer.MAX_VALUE);

        K = Integer.parseInt(br.readLine()); // 시작 정점
        edges = new ArrayList<>();
        for (int i=0; i<=V; i++) {
            edges.add(new ArrayList<int[]>());
        }
        for (int i=0; i<E; i++) {
            int u, v, w;
            st = new StringTokenizer(br.readLine());
            u = Integer.parseInt(st.nextToken());
            v = Integer.parseInt(st.nextToken());
            w = Integer.parseInt(st.nextToken());
            edges.get(u).add(new int[] {v, w});
        }

        dijkstra();

        for (int i=1; i<=V; i++) {
            System.out.println(dist[i] != Integer.MAX_VALUE ? dist[i] : "INF");
        }
    }

    static void dijkstra() {
        dist[K] = 0;

        pq = new PriorityQueue<>((o1, o2) -> {
            return Integer.compare(o1[1], o2[1]); // 가중치 오름차순 정렬
        });

        pq.add(new int[]{K, 0});

        while (!pq.isEmpty()) {
            int[] now = pq.poll();
            int v = now[0];
            int w = now[1];

            if (dist[v] < w) {
                continue;
            }

            for (int i=0; i<edges.get(v).size(); i++) {
                int[] next = edges.get(v).get(i);
                int nv = next[0];
                int nw = next[1];
                int newCost = w + nw;
                if (dist[nv] > newCost) {
                    dist[nv] = newCost;
                    pq.add(new int[] {nv, newCost});
                }
            }
        }
    }
}
